src/daemon/ubuntu_image_host.cpp

```diff
@@ -262,11 +267,38 @@ void mp::UbuntuVMImageHost::fetch_manifests()
         }
         catch (mp::DownloadException& e)
         {
+            if (is_force_update_from_network)
+            {
+                throw e;
+            }
             on_manifest_update_failure(e.what());
         }
         catch (const mp::UnsupportedRemoteException&)
         {
-            continue;
+        }
+    };
+
+    std::vector<std::future<void>> empty_futures;
+    empty_futures.reserve(remotes.size());
+    int index = 0;
+    for (const auto& [remote_name, remote_info] : remotes)
+    {
+        empty_futures.emplace_back(
+            std::async(std::launch::async, fetch_one_remote_and_write_to_index, index, remote_name, remote_info));
+        ++index;
+    }
+
+    for (auto& empty_future : empty_futures)
+    {
+        empty_future.get(); // use get instead of wait to retain the exception throwing
+    }
```

Comment:
This really seems like repeated code from earlier like in `custom_image_host.cpp` and `daemon.cpp`.  Any way to refactor this into a common way of doing this to avoid the code duplication?

Reply:
I was aware of that. I have given some thoughts on the `parallel_for` thing, but did not find a good way to do it, I can have a look again.